trin-core/src/lib.rs

```diff
@@ -6,3 +6,137 @@ pub mod jsonrpc;
 pub mod portalnet;
 pub mod socket;
 pub mod utils;
+
+pub const ACQUIRE_TIMEOUT_MS: u64 = 100;
+pub const HOLD_TIMEOUT_MS: u64 = 100;
+
+pub struct TimedRwReadGuard<'a, T: ?Sized> {
+    pub inner: tokio::sync::RwLockReadGuard<'a, T>,
+    pub acquisition_line: u32,
+    pub acquisition_file: &'static str,
+    pub acquisition_time: std::time::Instant,
+    pub sleep_task: tokio::task::JoinHandle<()>,
+}
+
+async fn sleep_then_log(file: &'static str, line: u32) {
+    tokio::time::sleep(std::time::Duration::from_millis(HOLD_TIMEOUT_MS)).await;
+    log::warn!(
+        "[{}:{}] lock held for over {}ms, not yet released",
+        file,
+        line,
+        HOLD_TIMEOUT_MS.to_string()
+    );
+}
+
+impl<'a, T: ?Sized> TimedRwReadGuard<'a, T> {
+    pub async fn new(
+        inner: tokio::sync::RwLockReadGuard<'a, T>,
+        acquisition_line: u32,
+        acquisition_file: &'static str,
+    ) -> TimedRwReadGuard<'a, T> {
+        let now = std::time::Instant::now();
+        let move_line = acquisition_line;
+        let move_file = acquisition_file;
+        let handle = tokio::spawn(async move {
+            sleep_then_log(move_file, move_line).await;
+        });
+
+        TimedRwReadGuard {
+            inner,
+            acquisition_line,
+            acquisition_file,
+            acquisition_time: now,
+            sleep_task: handle,
+        }
+    }
+}
+
+impl<'a, T> std::ops::Deref for TimedRwReadGuard<'a, T> {
+    type Target = tokio::sync::RwLockReadGuard<'a, T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'a, T: ?Sized> Drop for TimedRwReadGuard<'a, T> {
+    fn drop(&mut self) {
+        self.sleep_task.abort();
+        let held_for = self.acquisition_time.elapsed().as_millis();
+        if held_for > HOLD_TIMEOUT_MS.into() {
+            log::warn!(
+                "[{}:{}] lock held for too long: {}ms",
+                self.acquisition_file,
+                self.acquisition_line,
+                held_for,
+            )
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! rw_read {
+    ($lock_name:expr) => {{
+        let acquire_timeout = std::time::Duration::from_millis($crate::ACQUIRE_TIMEOUT_MS);
+        let sleep = tokio::time::sleep(acquire_timeout);
+        tokio::pin!(sleep);
+
+        let mut did_log: bool = false;
+        let now = std::time::Instant::now();
+
+        loop {
+            tokio::select! {
```

Comment:
Since there's already a dependency on `futures`, might be more understandable to write this using [`futures::future::select`](https://docs.rs/futures/0.3.17/futures/future/fn.select.html)?

Reply:
I'll give it a try

Reply:
I find the function based approach easier to read, though I don't really have that strong of an opinion on the subject.

Reply:
I wasn't able to find much of a difference between the two `select!`s but refactored this code a bit by unrolling the loop which should make this code a lot more understandable for first time readers.